File: semestre2/mpc/homeworks/conjunto-das-partes.py
from itertools import chain, combinations

def conjunto_das_partes(A):
    # Função para gerar todas as combinações possíveis dos elementos de A
    return list(chain.from_iterable(combinations(A, r) for r in range(len(A)+1)))

# Exemplo de uso
A = [1, 2, 3, 4]
resultado = conjunto_das_partes(A)

# Imprime o conjunto das partes
for subconjunto in resultado:
    print(subconjunto)

print('Tamanho do conjunto: ', subconjunto.__len__())
print("Possibilidades: ", resultado.__len__())

File: semestre2/mpc/homeworks/conjunto-das-partes.c
#include <stdio.h>
#include <math.h>

// Função para imprimir o conjunto das partes
void printPowerSet(int *set, int set_size) {
    unsigned int pow_set_size = pow(2, set_size);
    int counter, j;

    // Loop para gerar todas as combinações possíveis
    for (counter = 0; counter < pow_set_size; counter++) {
        printf("{ ");
        for (j = 0; j < set_size; j++) {
            // Verifica se o j-ésimo bit em counter está definido
            if (counter & (1 << j))
                printf("%d ", set[j]);
        }
        printf("}\n");
    }
}

int main() {
    int set[] = {1, 2, 3, 4, 5, 6};
    int set_size = sizeof(set) / sizeof(set[0]);

    printPowerSet(set, set_size);

    return 0;
}
